File: RecyclerViewJava/app/src/main/java/com/devwilly/recyclerview_java/RecyclerViewAdapter.java
package com.devwilly.recyclerview_java;

import com.devwilly.recyclerview_java.viewholders.BaseViewHolder;
import com.devwilly.recyclerview_java.viewholders.FoodViewHolder;
import com.devwilly.recyclerview_java.viewholders.FruitViewHolder;
import com.devwilly.recyclerview_java.viewholders.UnknownViewHolder;
import com.devwilly.recyclerview_java.wrapper.IEatingWrapper;

import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import java.util.ArrayList;


/**
 * Created by Willy on 15/01/2018.
 */

public class RecyclerViewAdapter extends RecyclerView.Adapter<BaseViewHolder> {

    public interface OnItemClickListener {
        void onFoodItemClick(String itemName);
    }

    private ArrayList<IEatingWrapper> mItemList = new ArrayList<>();
    private OnItemClickListener mOnItemClickListener;

    public RecyclerViewAdapter(ArrayList<IEatingWrapper> itemList) {
        this.mItemList = itemList;
    }

    public void setOnItemClickListener(OnItemClickListener listener) {
        this.mOnItemClickListener = listener;
    }

    @Override
    public BaseViewHolder onCreateViewHolder(ViewGroup parent, int viewType) {
        View view;
        LayoutInflater inflater = LayoutInflater.from(parent.getContext());

        switch (viewType) {
            case IEatingWrapper.FRUIT:
                view = inflater.inflate(R.layout.fruit_item_view, parent, false);
                return new FruitViewHolder(view);
            case IEatingWrapper.FOOD:
                view = inflater.inflate(R.layout.food_item_view, parent, false);
                return new FoodViewHolder(view, mOnItemClickListener);
            default:
                view = inflater.inflate(R.layout.unknown_item_view, parent, false);
                return new UnknownViewHolder(view);
        }
    }

    @Override
    public void onBindViewHolder(BaseViewHolder holder, int position) {
        holder.bindView(mItemList.get(position));
    }

    @Override
    public int getItemViewType(int position) {
        return mItemList.get(position).getItemType();
    }

    @Override
    public int getItemCount() {
        return mItemList.size();
    }
}

File: RecyclerView-Kotlin/app/src/main/java/com/devwilly/recyclerviewkotlin/RecyclerViewAdapter.kt
package com.devwilly.recyclerviewkotlin

import android.support.v7.widget.RecyclerView
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import com.devwilly.recyclerviewkotlin.viewholders.BaseViewHolder
import com.devwilly.recyclerviewkotlin.viewholders.FoodViewHolder
import com.devwilly.recyclerviewkotlin.viewholders.FruitViewHolder
import com.devwilly.recyclerviewkotlin.viewholders.UnknownViewHolder
import com.devwilly.recyclerviewkotlin.wrapper.IEatingWrapper

/**
 * Created by Willy on 15/01/2018.
 */
class RecyclerViewAdapter(itemList: ArrayList<IEatingWrapper>) : RecyclerView.Adapter<BaseViewHolder<*>>() {

    private var mItemList = ArrayList<IEatingWrapper>()

    init {
        mItemList = itemList
    }

    override fun onCreateViewHolder(parent: ViewGroup?, viewType: Int): BaseViewHolder<*> {
        val view: View
        val inflater: LayoutInflater = LayoutInflater.from(parent?.context)

        return when (viewType) {
            IEatingWrapper.FRUIT -> {
                view = inflater.inflate(R.layout.fruit_item_view, parent, false)
                FruitViewHolder(view)
            }
            IEatingWrapper.FOOD -> {
                view = inflater.inflate(R.layout.food_item_view, parent, false)
                FoodViewHolder(view)
            }
            else -> {
                view = inflater.inflate(R.layout.unknown_item_view, parent, false)
                UnknownViewHolder(view)
            }
        }


    }

    override fun onBindViewHolder(holder: BaseViewHolder<*>, position: Int) {
        when (holder) {
            is FruitViewHolder -> holder.bindView(mItemList[position])
            is FoodViewHolder -> holder.bindView(mItemList[position])
            is UnknownViewHolder -> holder.bindView(mItemList[position])
        }
    }

    override fun getItemViewType(position: Int): Int {
        return mItemList[position].getItemType()
    }

    override fun getItemCount(): Int {
        return mItemList.size
    }
}